### ISSUE TITLE
setState being invoked when removing CustomLayer

### ISSUE THREAD

**Issue by junhanamaki** [gaearon/react-dnd#307]
Hello.

My use case requires for the component being dragged to stay at the same place, and at the same time show a different image being dragged. Because of issues with css styles I would like to achieve this by rendering a copy of the original component inside the same component, and dragging it around by applying transforms with position fixed. To achieve this I though of using a CustomDragLayer that gets rendered when the component is dragging, meaning that this CustomDragLayer would be dynamically added and removed. The problem is that I'm getting "Warning: setState(...): Can only update a mounted or mounting component. This usually means you called setState() on an unmounted component. This is a no-op. Please check the code for the DragLayer(Component) component.". A sample code as follows:

```javascript
const { DragDropContext, DragSource, DragLayer } = require('react-dnd');
const HTML5Backend = require('react-dnd/modules/backends/HTML5');

const TestBoard = DragDropContext(HTML5Backend)(React.createClass({
  render() {
    return <div><Draggable /></div>;
  }
}));

const Layer = (() => {
  function collect(monitor) {
    return {
      // removing this line will make it work
      initialOffset: monitor.getInitialSourceClientOffset()
    };
  }

  const Component = React.createClass({
    render() {
      return <div>I am... the CustomDragLayer!</div>;
    }
  });

  return DragLayer(collect)(Component);
})();

const Draggable = (() => {
  function collect(connect, monitor) {
    return {
      connectDragSource: connect.dragSource(),
      isDragging: monitor.isDragging()
    };
  }

  const source = {
    beginDrag: () => ({})
  };

  const Component = React.createClass({
    render() {
      return (
        <div>
          {this.props.connectDragSource(<div>drag me please!</div>)}
          {this.renderLayer()}
        </div>
      );
    },

    renderLayer() {
      if (this.props.isDragging) {
        return <Layer />;
      }
    }
  });

  return DragSource('TestType', source, collect)(Component);
})();

ReactDOM.render(<TestBoard />, document.body);
```

The issue seems to be in the collect function for DragLayer, where I call "monitor.getInitialSourceClientOffset()", removing this will remove the warning. Going a little bit further it seems that the unsubscribe method of redux store is invoked during dispatch, meaning that it will invoke the listener for CustomDragLayer, even though it was already unsubscribed.

Sorry for the long post and thank you for your attention, hope it makes sense :)

**Comment by gaearon** [comment]
Do you know how to fix this?

**Comment by junhanamaki** [comment]
Hello.

Unfortunately I'm new to react and started to use dnd (and by extension redux) only recently, so no, I don't know how to fix this in a valid way.

Just for the sake of discussion I could suggest during the dispatch method in redux check if the listener method is still in the original list before invoking, or in dnd check if component is still mounted before doing setState (even though this.isMounted is deprecated), but I still don't understand enough of both to evaluate if these are valid fixes (or if any of them even need fixes to begin with).

Sorry for not being of much help.

**Comment by gaearon** [comment]
Can you provide a repo with minimal example reproducing this problem?

**Comment by gaearon** [comment]
Oh, I see your example is actually self-contained.
Let me get back to this when I have some more time to work on DnD.
(Unfortunately not likely to be this week.)

**Comment by gaearon** [comment]
Just to clarify—it isn't throwing, is it? You're just seeing a warning.

**Comment by junhanamaki** [comment]
Yes, it's just a warning.

**Comment by junhanamaki** [comment]
Also thanks for your attention.

**Comment by junhanamaki** [comment]
Hi again.

I think I finally understood why this happens, it's because DragLayer's collect function returns a value which is different from the previous when doing monitor.getInitialSourceClientOffset(), thus triggering the setState in the DragLayer's handleChange, so what seems to be happening is:

1 - stop dragging, this triggers an action of type "dnd-core/END_DRAG" to redux's dispatch;
2 - dispatch invokes listeners by creating a copy of the list;
3 - while invoking the listeners the custom drag layer stops being rendered, thus being unmounted;
4 - custom drag layer unsubscribes listener in componentwillunmount;
5 - since dispatch is using the original list of listeners it will still have reference to the removed listener, thus invoking it;
6 - when invoked, because the value returned from the collect function is different from the previous value (because of monitor.getInitialSourceClientOffset()) it will trigger setState.

The code portions in question:

```javascript
// react-dnd, file DragLayer.js: shallowEqual is the reason why it happens only when doing monitor.getInitialSourceClientOffset()
handleChange() {
  const nextState = this.getCurrentState();
  if (!shallowEqual(nextState, this.state)) {
    this.setState(nextState);
  }
}
```

```javascript
// redux, file createStore.js, function dispatch: listeners removed during iteration will still be called
listeners.slice().forEach(function (listener) {
  return listener();
});
```

Wouldn't it be more reliable and predictable if in redux, a removed listener is not called at all, or at least be called with a flag that indicates that it was removed (it could be useful for clean up logic)?

Hope I'm not annoying too much...

Thank you for your attention.

**Comment by gaearon** [comment]
Should be fixed in 2.1.0.
